src/connectivity/percolation.c

```diff
@@ -0,0 +1,281 @@
+        // Use random edge order
+        generated_edges = true;
+        igraph_integer_t size = igraph_ecount(graph);
+
+        IGRAPH_CHECK(igraph_vector_int_init_range(&new_edges, 0, size));
+
+        IGRAPH_CHECK(igraph_vector_int_shuffle(&new_edges));
+
+        IGRAPH_FINALLY(igraph_vector_int_destroy, &new_edges);
+        internal_edge_indices = &new_edges;
+    }
+
+    igraph_vector_int_t edgelist;
+    IGRAPH_CHECK(igraph_vector_int_init(&edgelist, 2 * igraph_vector_int_size(internal_edge_indices)));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &edgelist);
+
+    for (igraph_integer_t i = 0; i < igraph_vector_int_size(internal_edge_indices); i++) {
+        IGRAPH_CHECK(igraph_edge(graph,
+                                 VECTOR(*internal_edge_indices)[i],
+                                 &VECTOR(edgelist)[2 * i],
+                                 &VECTOR(edgelist)[2 * i + 1]));
+    }
+
+
+    IGRAPH_CHECK(edge_list_percolation(&edgelist, output));
+
+    igraph_vector_int_destroy(&edgelist);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    if (generated_edges) {
+        igraph_vector_int_destroy(&new_edges);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t percolate_site(const igraph_t *graph,
+                                     igraph_vector_int_t *links,
+                                     igraph_vector_int_t *sizes,
+                                     igraph_integer_t *biggest,
+                                     igraph_integer_t vertex) {
+
+    if (VECTOR(*sizes)[vertex] != 0) {
+        IGRAPH_ERROR("Vertex already added.", IGRAPH_EINVAL);
+    }
+
+    VECTOR(*sizes)[vertex] = 1;
+
+    igraph_vector_int_t neighbors;
+    IGRAPH_CHECK(igraph_vector_int_init(&neighbors, 0));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &neighbors);
+
+    IGRAPH_CHECK(igraph_neighbors(graph, &neighbors, vertex, IGRAPH_IN));
+
+    for (igraph_integer_t i = 0; i < igraph_vector_int_size(&neighbors); i++) {
+        if (VECTOR(*sizes)[VECTOR(neighbors)[i]] == 0) {
+            continue;
+        }
+        percolate_edge(links, sizes, biggest, vertex, VECTOR(neighbors)[i]);
+    }
+
+    igraph_vector_int_destroy(&neighbors);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+igraph_error_t igraph_site_percolation(
+        const igraph_t *graph,
+        igraph_vector_int_t *output,
+        igraph_vector_int_t *vertices) {
+
+    // If vertex list is null, generate a random one.
+    igraph_vector_int_t* internal_vertices;
+    igraph_bool_t generated_vertices = false;
+    igraph_vector_int_t new_vertices;
+
+    if (vertices == NULL) {
+        // Use random vertex order
+        generated_vertices = true;
+        igraph_integer_t size = igraph_vcount(graph);
+
+        IGRAPH_CHECK(igraph_vector_int_init_range(&new_vertices, 0, size));
+
+        IGRAPH_CHECK(igraph_vector_int_shuffle(&new_vertices));
+
+        IGRAPH_FINALLY(igraph_vector_int_destroy, &new_vertices);
```

Comment:
```suggestion
        IGRAPH_CHECK(igraph_vector_int_init_range(&new_vertices, 0, size));
        IGRAPH_FINALLY(igraph_vector_int_destroy, &new_vertices);
        IGRAPH_CHECK(igraph_vector_int_shuffle(&new_vertices));
```

See the analogous comment above.